src/VisualStudio/Core/Impl/CodeModel/ProjectCodeModelFactory.cs

```diff
@@ -199,15 +182,14 @@ public EnvDTE.FileCodeModel GetOrCreateFileCodeModel(ProjectId id, string filePa
             => GetProjectCodeModel(id).GetOrCreateFileCodeModel(filePath).Handle;
 
         public void ScheduleDeferredCleanupTask(Action a)
-            => _deferredCleanupTasks.Add(_threadingContext.JoinableTaskFactory.StartOnIdle(a, VsTaskRunContext.UIThreadNormalPriority));
-
-        void IDisposable.Dispose()
         {
-            // Stop any outstanding BG work we've queued up.
-            _cancellationTokenSource.Cancel();
-
-            // Now wait for all our cleanup tasks to finish before we return.
-            _deferredCleanupTasks.Join();
+            var joinableTask = _threadingContext.JoinableTaskFactory.StartOnIdle(a, VsTaskRunContext.UIThreadNormalPriority);
+            _threadingContext.ShutdownBlockingTasks.Add(joinableTask);
+            if (_threadingContext.DisposalToken.IsCancellationRequested)
+            {
+                // We are already in the process of shutting down. Wait for the deferred task to complete.
+                joinableTask.Join();
```

Comment:
jeez.  i know you're maintaining behavior.  but having us do allt his blocking stuff on shutdown just seems awful.  i'm not sure who could benefit from this.  It just seems like we're forcing blocking work on a path we want to be fast.

Reply:
Yeah this is really weird. _Really_ weird.

Reply:
i'm fine if you don't want to mess with this in this pr.  but a tracking bug woudl be nice here so we don't lose this.

Reply:
The only blocking operation we actually do after shutdown is requested is save a modified document, which seems fine.